File: .orkestrel/campaign/docs-parity/instruments/d7/units/form/d7n-form-converge/cells.py
import io, re, sys

SPLIT = re.compile(r'(?<!\\)\|')

def tables(path):
    with io.open(path, encoding='utf-8') as fh:
        lines = fh.read().split('\n')
    rows = {}
    i = 0
    while i < len(lines):
        line = lines[i]
        if line.startswith('|') and i + 1 < len(lines) and set(lines[i + 1].replace('|', '').replace(' ', '').replace(':', '')) == {'-'}:
            header = [c.strip() for c in SPLIT.split(line)[1:-1]]
            j = i + 2
            while j < len(lines) and lines[j].startswith('|'):
                cells = [c.strip() for c in SPLIT.split(lines[j])[1:-1]]
                key = (header[0], cells[0])
                rows[key] = dict(zip(header, cells))
                j += 1
            i = j
            continue
        i += 1
    return rows

before = tables(sys.argv[1])
after = tables(sys.argv[2])
missing = [k for k in before if k not in after]
added = [k for k in after if k not in before]
changed = []
for key, cells in before.items():
    if key not in after:
        continue
    for column, value in cells.items():
        if column == 'Summary':
            continue
        now = after[key].get(column)
        if now != value:
            changed.append((key, column, value, now))
print('rows before: %d, rows after: %d' % (len(before), len(after)))
print('rows missing after: %s' % (missing or 'none'))
print('rows added after: %s' % (added or 'none'))
print('non-Summary cells changed: %d' % len(changed))
for row in changed:
    print('  %s [%s]\n    before: %s\n    after:  %s' % row)

File: .orkestrel/campaign/docs-parity/instruments/d7/units/lsp/d7n-lsp-converge/cells.mjs
// Compares every non-`Summary` cell of every table row against the baseline guide.
// A row is keyed by its first cell; the inserted `Shape` cell is reported separately.
import { readFileSync } from 'node:fs'

/**
 * Splits one Markdown table row into its cells.
 *
 * @param line - The row, leading and trailing pipe included.
 * @returns The trimmed cells between the outer pipes.
 */
function splitRow(line) {
	const inner = line.trim().replace(/^\|/, '').replace(/\|$/, '')
	return inner.split(/(?<!\\)\|/).map((cell) => cell.trim())
}

/**
 * Indexes every table row of one guide by its first cell.
 *
 * @param path - The guide to read.
 * @returns Each row's key mapped to its cells.
 */
function readRows(path) {
	const rows = new Map()
	for (const line of readFileSync(path, 'utf8').split('\n')) {
		if (!line.startsWith('| ')) continue
		const cells = splitRow(line)
		if (/^-+$/.test(cells[0])) continue
		if (rows.has(cells[0])) continue
		rows.set(cells[0], cells)
	}
	return rows
}

const before = readRows(process.argv[2])
const after = readRows(process.argv[3])
let compared = 0
let mismatched = 0
let missing = 0
let shaped = 0
for (const [key, cells] of before) {
	const now = after.get(key)
	if (now === undefined) {
		missing += 1
		process.stdout.write(`missing after: ${key}\n`)
		continue
	}
	const widened = now.length === cells.length + 1
	if (widened && now[2].length > 0) shaped += 1
	const kept = widened ? [now[0], now[1], ...now.slice(3)] : now
	for (let index = 0; index < kept.length - 1; index += 1) {
		compared += 1
		if (kept[index] === cells[index]) continue
		mismatched += 1
		process.stdout.write(`changed ${key} cell ${index}: ${cells[index]} -> ${kept[index]}\n`)
	}
}
process.stdout.write(
	`rows compared: ${before.size}, non-Summary cells compared: ${compared}, mismatched: ${mismatched}, missing after: ${missing}, Shape cells filled: ${shaped}\n`,
)
